### PR TITLE
SnowflakeToS3Operator

### PR DESCRIPTION
**Pull request by JavierLopezT**
New transfer operator, Snowflake To S3

### CODE REVIEW

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (turbaszek) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: Missing Apache license header

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (turbaszek) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]:
```suggestion
        self,
        *,
        s3_bucket: str,
        s3_key: str,
        file_format: str,
        schema: Optional[str] = None,
        table: Optional[str] = None,
        sql: Optional[str] = None,
        snowflake_conn_id: str = "snowflake_default",
        unload_options: Optional[str] = None,
        autocommit: bool = True,
        **kwargs) -> None:
```

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (turbaszek) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: Missing indent

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (turbaszek) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: Missing indent

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (turbaszek) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: Do we need this `\n\t\t\t` because the `unload_query` is indented? Should we be able to dedent it?

Comment (JavierLopezT) [reply airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: To be honest, I just copied it from the existing RedshiftToS3Operator. I don't know if we should maintain this or modifying. What do you think?

Comment (turbaszek) [reply airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: I think if we can improve the code and remove this strange string then it may be worth doing it

Comment (JavierLopezT) [reply airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: I was able to remove two \t characters. Otherwise, the query in the log is unintended.

**Review by turbaszek — APPROVED** [review]

**Comment by ryw** [comment]
@turbaszek is this mergeable or waiting on another review?

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (RosterIn) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: I think this operator needs further documentation.
Some users might be surprised that there is no `aws_conn` here and the `aws_access_key_id`, `aws_secret_access_key` are defined in snowflake connection. I think this is not trivial.

File: tests/providers/amazon/aws/transfers/test_snowflake_to_s3.py
Comment (RosterIn) [inline tests/providers/amazon/aws/transfers/test_snowflake_to_s3.py]: The operator has two modes one when sql param is provided and one when it's not. Shouldn't be two test cases?

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (feluelle) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]:
```suggestion
```
I think this makes it unnecessary "complex".

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (feluelle) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]:
```suggestion
        table_or_query: Optional[str] = None,
```
I would suggest to keep it rather simple and allow passing table or query to one variable and just pass it to the `COPY` string. Otherwise you would have to do more validation on this - so that either schema and table or a query was passed.

Ref: https://docs.snowflake.com/en/sql-reference/sql/copy-into-location.html#syntax

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (JavierLopezT) [reply airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: I have added some docs as well as another possibility to unload. What do you think?

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (JavierLopezT) [reply airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: Done it.

**Comment by JavierLopezT** [comment]
I have added a new possibility to make the unload. Not sure if this is the more elegant way to do it, though. So before applying the changes in the test I would like some feedback. Thanks!

**Comment by kaxil** [comment]
Can you please rebase your PR on latest Master since we have applied [Black](https://github.com/apache/airflow/commit/4e8f9cc8d02b29c325b8a5a76b4837671bdf5f68) and [PyUpgrade](https://github.com/apache/airflow/commit/8c42cf1b00c90f0d7f11b8a3a455381de8e003c5) on Master.

It will help if your squash your commits into single commit first so that there are less conflicts.

**Comment by JavierLopezT** [comment]
> Can you please rebase your PR on latest Master since we have applied [Black](https://github.com/apache/airflow/commit/4e8f9cc8d02b29c325b8a5a76b4837671bdf5f68) and [PyUpgrade](https://github.com/apache/airflow/commit/8c42cf1b00c90f0d7f11b8a3a455381de8e003c5) on Master.
> 
> It will help if your squash your commits into single commit first so that there are less conflicts.

I have done it (without squashing though, I didn't know how to do it), but has it worked? I didn't see any conflict

**Comment by RosterIn** [comment]
@JavierLopezT the checks are failing with
`tests/providers/amazon/aws/transfers/test_snowflake_to_s3.py:43:8: E1125: Missing mandatory keyword argument 'query_or_table' in constructor call (missing-kwoa)`

**Comment by JavierLopezT** [comment]
> @JavierLopezT the checks are failing with
> `tests/providers/amazon/aws/transfers/test_snowflake_to_s3.py:43:8: E1125: Missing mandatory keyword argument 'query_or_table' in constructor call (missing-kwoa)`

Thanks @RosterIn . That is because I didn't update the code in the tests to match the operator, because as I said in https://github.com/apache/airflow/pull/11332#issuecomment-720435299 I wanted some feedback before proceeding. Probably my mistake to not tag anyone. If you or @feluelle don't have any objections, I'll fix the tests ASAP

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (RosterIn) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]:
```suggestion
        `schema.table`
```

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (feluelle) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]:
```suggestion
    :type snowflake_conn_id: str
```

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (feluelle) [inline airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]:
```suggestion
    :param use_stage: Copy the data into a Snowflake Stage. This allows you to use the method
        'Configuring AWS IAM User Credentials' for unloading data to s3. If this is true,
        s3_bucket and file_format are not necessaries.
    :type use_stage: bool
```
WDYT? If yes, then obviously you have to change the references. (I do not comment on them. :))

**Review by feluelle — COMMENTED** [review]
LGTM, only one suggestion :) 👍

File: airflow/providers/amazon/aws/transfers/snowflake_to_s3.py
Comment (JavierLopezT) [reply airflow/providers/amazon/aws/transfers/snowflake_to_s3.py]: Actually this is a mistake, stage is a string. I have already changed it. Thanks!

**Review by feluelle — APPROVED** [review]

**Comment by feluelle** [comment]
@JavierLopezT The static check is failing because of:
```
Validate providers.yaml files............................................................Failed
- hook id: provider-yamls
- exit code: 1

Checking integration duplicates
Checking completeness of list of {sensors, hooks, operators}
Checking for duplicates in list of {sensors, hooks, operators}
Checking completeness of list of transfers
Checking for duplicates in list of transfers
Checking connection classes belong to package
Checking doc files
Detect unregistered integrations
Found 2 errors
The `airflow.providers.aws.transfers.snowflake_to_s3` object in transfers list in airflow/providers/amazon/provider.yaml does not start with the expected airflow.providers.amazon.

Incorrect content of key 'transfers/python-module' in file: airflow/providers/amazon/provider.yaml
  Items in the first set but not the second:
  'airflow.providers.amazon.aws.transfers.snowflake_to_s3'
  Items in the second set but not the first:
  'airflow.providers.aws.transfers.snowflake_to_s3'
```

**Pull request closed without merging**
